infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -81,19 +90,74 @@
             .getSerializedSize();
     private final boolean noVerify;
     public final String logDir;
-    private Map<String, SegmentHandle> writeChannels;
+    private LoadingCache<Pair<UUID, Long>, SegmentHandle> segmentsCache;
     private Set<FileChannel> channelsToSync;
     private MultiReadWriteLock segmentLocks = new MultiReadWriteLock();
 
     public StreamLogFiles(String logDir, boolean noVerify) {
+        this(logDir, noVerify, DEFAULT_SEGMENT_CACHE_SIZE);
+    }
+
+    public StreamLogFiles(String logDir, boolean noVerify, long segmentCacheSize) {
         this.logDir = logDir;
-        writeChannels = new ConcurrentHashMap();
+        segmentsCache = getCache(segmentCacheSize);
         channelsToSync = new HashSet<>();
         this.noVerify = noVerify;
 
         verifyLogs();
     }
 
+    @SuppressWarnings("unchecked")
+    private LoadingCache<Pair<UUID, Long>, SegmentHandle> getCache(long numSegments) {
+        return CacheBuilder.newBuilder()
+                .maximumWeight(numSegments)
+                .weigher((Long, SegmentHandle) -> {
+                            return 1;
+                        }
+                )
+                .removalListener(new RemovalListener() {
+                    @Override
+                    public void onRemoval(RemovalNotification notification) {
+                        // Would this cause NPE for other objects holding this segment's reference?
```

Comment:
The comment was actually not about an NPE inside the removal listener, it was about other objects holding references to the channel inside the segment handler after it has been evicted from the cache. For example fh.logChannel.write(recordBuf); in writeRecord will throw an NPE if fh (segment handle) has been evicted within the critical time described above,